script/run-unit-test.js

```diff
@@ -44,12 +51,98 @@ const testRunner = options.coverage ? coveragePath : mochaPath;
 
-// Otherwise, run the command
-runCommand(
-  `LOG_LEVEL=${options[
-    'log-level'
-  ].toLowerCase()} ${testRunner} --max-old-space-size=4096 --opts ${mochaOpts} --recursive ${options.path
-    .map(p => `'${p}'`)
-    .join(' ')}`,
-  options.coverage ? null : 0,
-);
+const logLevel = options['log-level'].toLowerCase();
+const forcedExitCode = options.coverage ? null : 0;
+const testDirectories = options.path;
+
+// run all unit tests else run only the test provided as a command line argument
+if (testDirectories[0] === defaultPath) {
+  // array of unit test paths
+  const allUnitTestPaths = glob.sync(defaultPath);
+
+  // Group unit tests into suites by directories by creating a list of all unit test containing directories from the list of allUnitTestPaths
+  const unitTestDirectories = new Set(
+    allUnitTestPaths.map(unitTest => {
+      // Start with './src/applications/burials/tests/config/benefitsSelection.unit.spec.js'
+      const directory = path.dirname(unitTest);
+      const directoryPathArray = directory.split(path.sep);
+
+      let subdirectoryDepth = 3;
+
+      if (directoryPathArray[2] === 'applications') {
+        subdirectoryDepth = 4;
+
+        if (directoryPathArray[3] === 'personalization') {
+          subdirectoryDepth = 5;
+        }
+      }
+
+      // Reduce to 'src/applications/burials/'
+      return directoryPathArray.slice(1, subdirectoryDepth).join(path.sep);
+    }),
+  );
+
+  // When the "coverage" arg is passed to this script, we execute Mocha with the
+  // "reporter" arg set to use JUnit reports. This results in an XML file being
+  // written to the file system after each Mocha execution.
+  let junitReportForAllTestRuns = null;
+  const junitReportFilePath = path.join(__dirname, '../test-results.xml');
+
+  const wildCard = `${path.sep}**${path.sep}*.unit.spec.js?(x)`;
+  const failures = [];
+
+  unitTestDirectories.forEach(unitTestDirectory => {
+    // eslint-disable-next-line no-console
+    console.log(chalk.blue(`Test suite: ${unitTestDirectory}`));
+
+    let executeMocha = null;
+
+    if (options.coverage) {
+      executeMocha = `NODE_ENV=test nyc --include "${unitTestDirectory}/**" --reporter=lcov --reporter=text --reporter=json-summary mocha --reporter mocha-junit-reporter --no-color`;
+    } else {
+      executeMocha = 'BABEL_ENV=test mocha';
+    }
+
+    const findUnitTestsInDir = `".${path.sep}${unitTestDirectory}${wildCard}"`;
+    const command = `LOG_LEVEL=${logLevel} ${executeMocha} --opts ${mochaOpts} --recursive ${findUnitTestsInDir}`;
+    const exitStatus = runCommandSync(command);
+
+    if (options.coverage) {
+      // A JUnit Report was just written to the file system in the form of a file "test-results.xml".
+      // We need to read that report into memory and concatenate its contents into a comprehensive JUnit Report
+      // that contains the result of all test runs.
+
+      const junitReport = cheerio.load(
```

Comment:
Isn't junit a java testing framework? I'm a little confused why this is needed (I didn't even know we were using junit lol) 🙂

Reply:
I was just reading the build arg passed to Mocha. I think it's just an output format for test runs that Jenkins can read.